nmrfx-analyst-gui/src/main/java/org/nmrfx/processor/gui/FXMLControllerManager.java

```diff
@@ -0,0 +1,163 @@
+    private final SimpleObjectProperty<FXMLController> activeController = new SimpleObjectProperty<>(null);
+
+    // A collection of all known controllers.
+    // It is important to keep insertion order: the first controller is considered the main one and is kept when closing all secondary stages
+    private final Set<FXMLController> controllers = new LinkedHashSet<>();
+
+    /**
+     * Get all registered controllers, keeping the insertion order.
+     *
+     * @return an unmodifiable collection of registered controllers.
+     */
+    public Collection<FXMLController> getControllers() {
+        return Collections.unmodifiableSet(controllers);
+    }
+
+    /**
+     * Register a new controller and mark it as active.
+     *
+     * @param controller the controller to register
+     */
+    private void registerNewController(FXMLController controller) {
+        controllers.add(controller);
+        activeController.set(controller);
+    }
+
+    /**
+     * Unregister a controller. To be called when the associated stage is closed.
+     *
+     * @param controller the controller to unregister.
+     */
+    public void unregister(FXMLController controller) {
+        controllers.remove(controller);
+    }
+
+    /**
+     * Check if a particular controller is already registered.
+     *
+     * @return true if the controller is known.
+     */
+    public boolean isRegistered(@Nullable FXMLController controller) {
+        return controller != null && controllers.contains(controller);
+    }
+
+    /**
+     * Get the active controller, creating one if necessary.
+     *
+     * @return the active controller.
+     */
+    @Nonnull
+    public FXMLController getOrCreateActiveController() {
+        FXMLController active = activeController.get();
+        if (active == null) {
+            //XXX is there actually a case where not having an active controller is a possibility?
+            active = newController();
+
+            //XXX in which case could we not already be the active controller here?
+            //XXX it looks like creating a controller already sets it as active
+            setActiveController(active);
+        }
+
+        return active;
+    }
+
+    /**
+     * Exposes the active controller as a property, so other components can use it as an event source.
+     *
+     * @return the active controller property.
+     */
+    public SimpleObjectProperty<FXMLController> activeControllerProperty() {
+        return activeController;
+    }
+
+    /**
+     * Set the active controller or null.
+     *
+     * @param active the active controller.
+     * @throws IllegalStateException if the controller is not registered
+     */
+    public void setActiveController(@Nullable FXMLController active) {
+        if (active != null && !controllers.contains(active)) {
+            throw new IllegalStateException("Trying to set an unregistered controller as the active one!");
+        }
```

Comment:
Encountered this exception following these steps
1. Start NMRFx
2. Go to Spectra -> New Window
3. Close original window
4. Go to Projects -> Save and save project
5. Check logs for exception

Reply:
Found and fixed the root cause: GUIScripter keeps a FXMLController instance in a static field, but isn't notified when the window is closed and unregistered from the instance list.

This causes a temporary memory leak (the controller should be garbaged but will not be until another one is set to GUIScripter) and the bug you found. I've kept the static field but replaced by the active one when calling `GUIScripter.getControlller()` if it has been unregistered.

I've kept the exception in place in FXMLController manager since I believe it can help us find other bugs similar to this one.